src/ddmd/attrib.d

```diff
@@ -1349,6 +1349,140 @@ extern (C++) final class StaticIfDeclaration : ConditionalDeclaration
     }
 }
 
+extern (C++) final class StaticForeachDeclaration : AttribDeclaration
+{
+    StaticForeach sfe;
+
+    ScopeDsymbol scopesym;
+    bool addisdone;
+
+    Dsymbols* cache = null;
+
+    extern (D) this(StaticForeach sfe, Dsymbols* decl)
+    {
+        super(decl);
+        this.sfe = sfe;
+    }
+
+    override Dsymbol syntaxCopy(Dsymbol s)
+    {
+        assert(!s);
+        return new StaticForeachDeclaration(
+            sfe.syntaxCopy(),
+            Dsymbol.arraySyntaxCopy(decl));
+    }
+
+    override final bool oneMember(Dsymbol* ps, Identifier ident)
+    {
+        // TODO: is there a better implementation for this?
+        *ps = null;
+        return true;
+    }
+
+    override Dsymbols* include(Scope* sc, ScopeDsymbol sds)
+    {
+        if (cache)
+        {
+            return cache;
+        }
+        sfe.prepare(_scope);
+        if (!sfe.ready())
+        {
+            return null; // TODO: ok?
+        }
+        import ddmd.statementsem: makeTupleForeach;
+        Dsymbols* d = makeTupleForeach!(true,true)(_scope, sfe.aggrfe, decl, sfe.needExpansion);
+        if (d && !addisdone)
+        {
+            // Add members lazily.
+            for (size_t i = 0; i < d.dim; i++)
+            {
+                Dsymbol s = (*d)[i];
+                s.addMember(_scope, scopesym);
+            }
+            // Set the member scopes lazily.
+            for (size_t i = 0; i < d.dim; i++)
+            {
+                Dsymbol s = (*d)[i];
+                s.setScope(_scope);
+            }
+            addisdone = true;
+        }
+        cache = d;
+        return d;
+    }
+
+    override void addMember(Scope* sc, ScopeDsymbol sds)
+    {
+        this.scopesym = sds;
+    }
+
+    override final void addComment(const(char)* comment)
+    {
+        // TODO
+    }
+
+    override void setScope(Scope* sc)
+    {
+        // do not evaluate condition before semantic pass
+        // But do set the scope, in case we need it for forward referencing
+        Dsymbol.setScope(sc);
+    }
+
+    override void importAll(Scope* sc)
+    {
+        // do not evaluate aggregate before semantic pass
+    }
+
+    override void semantic(Scope* sc)
+    {
+        AttribDeclaration.semantic(sc);
+    }
+
+    override const(char)* kind() const
+    {
+        return "static foreach";
+    }
+
+    override void accept(Visitor v)
+    {
+        v.visit(this);
+    }
+}
+
+extern(C++) final class ForwardingAttribDeclaration: AttribDeclaration
```

Comment:
What is this for? Perhaps add a small paragraph for each new class describing their role - maybe even including a code sample.